File: 046. Permutations/permute.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
    var dfs = (nums,path,res) => {
        if (nums.length==0) {
            res.push([...path]) //must [...path]
            return
        }
        else {
            const len = nums.length
            for (let i=0;i<len;i++) {
            // for (let num of nums) {
                path.push(nums[i])
                new_nums = nums.slice(0,i).concat(nums.slice(i+1))
                // new_nums = nums.filter(item => item !== nums[i])
                dfs(new_nums,path,res)
                path.pop()
            }
        }
    }
    
    res = []
    dfs(nums,[],res)
    return res
};

// Runtime 125 ms / Memory 44.7 MB
// debug: https://jsfiddle.net/

File: 046. Permutations/permute.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> path;
        vector<int> visited(nums.size(), 0);
        DFS(nums, visited, path, res);
        return res;
    }

    void DFS(vector<int> &nums, vector<int> &visited, vector<int> &path, vector<vector<int>> &res) {
        // full and return
        if (path.size() == nums.size()) {
            res.push_back(path);
            return;
        }
        
        for (int i=0; i<nums.size(); i++) {
            // already used
            if (visited[i] == 1) { continue; }
            // DFS
            else {
                path.push_back(nums[i]);
                visited[i] = 1;
                DFS(nums, visited, path, res);
                path.pop_back();
                visited[i] = 0;
            }
        }
    }
};